polygeist/lib/Dialect/Polygeist/Transforms/FunctionSpecialization.cpp

```diff
@@ -0,0 +1,281 @@
+//     %arg1 : memref<?x!sycl.accessor) {}
+//   gpu.func @caller kernel {
+//     func.call @callee(%acc1, %acc2)
+//   }
+// Optimized code:
+//   func.func private @callee.specialized(
+//     %arg0 : memref<?x!sycl.accessor> {sycl.inner.disjoint},
+//     %arg1 : memref<?x!sycl.accessor> {sycl.inner.disjoint}) {}
+//   func.func private @callee(
+//     %arg0 : memref<?x!sycl.accessor>,
+//     %arg1 : memref<?x!sycl.accessor) {}
+//   gpu.func @caller kernel {
+//     if (not_overlap(%acc1, %acc2))
+//       func.call @callee.specialized(%acc1, %acc2)
+//     else
+//       func.call @callee(%acc1, %acc2)
+//   }
+class FunctionSpecializationPass
+    : public polygeist::impl::FunctionSpecializationBase<
+          FunctionSpecializationPass> {
+public:
+  using FunctionSpecializationBase<
+      FunctionSpecializationPass>::FunctionSpecializationBase;
+
+  void runOnOperation() override;
+
+private:
+  /// Returns true if \p func is a candidate.
+  bool isCandidateFunction(FunctionOpInterface func) const;
+  /// Returns true if \p acc1 and \p acc2 need to be checked for no overlap. For
+  /// example, under strict aliasing rule, accessors with different element
+  /// types are not alias, so return false.
+  bool isCandidateAccessorPair(AccessorPtrType acc1,
+                               AccessorPtrType acc2) const;
+  /// Populate \p accessorPairs with accessor pairs that should be checked for
+  /// no overlap for \p call.
+  void collectMayOverlapAccessorPairs(
+      CallOpInterface call,
+      SmallVectorImpl<AccessorPtrPairType> &accessorPairs) const;
+  /// Version \p call.
+  void versionCall(CallOpInterface call) const;
+};
+
+} // namespace
+
+//===----------------------------------------------------------------------===//
+// FunctionSpecializationPass
+//===----------------------------------------------------------------------===//
+
+void FunctionSpecializationPass::runOnOperation() {
+  SmallVector<FunctionOpInterface> candidates;
+  getOperation()->walk([&](FunctionOpInterface func) {
+    LLVM_DEBUG(llvm::dbgs()
+               << "Processing function \"" << func.getName() << "\"\n");
+    if (isCandidateFunction(func))
+      candidates.push_back(func);
+  });
+
+  ModuleOp module = getOperation();
+  SymbolTableCollection symTable;
+  SymbolUserMap userMap(symTable, module);
+
+  for (FunctionOpInterface func : candidates) {
+    FunctionOpInterface clonedFunc = cloneFunction(func);
+    setInnerDisjointAttribute(clonedFunc, isCandidateArg);
+
+    for (Operation *op : userMap.getUsers(func)) {
+      // Due to temporary condition to only allow function called directly by a
+      // GPU kernel.
+      assert(op->getParentOfType<gpu::GPUFuncOp>() &&
+             "Expecting calls only in GPU kernel");
+
+      auto call = cast<CallOpInterface>(op);
+      versionCall(call);
+      updateCallee(call, clonedFunc);
+    }
+    ++numFunctionSpecialized;
+  }
+}
+
+bool FunctionSpecializationPass::isCandidateFunction(
```

Comment:
Maybe we can enforce here that every caller is a `func.call`.

Reply:
Let's continue the discussion in one thread, https://github.com/intel/llvm/pull/9131#discussion_r1173923377.